Rework the "by store" X509_LOOKUP method to open the given URI early

levitte opened the pull request:
The cached `X509_LOOKUP` method data is no longer just the URI, but now
includes the `OSSL_STORE_CTX` pointer, and required parameters to reopen
the URI at any time.  `cache_objects()` is modified to handle this, and
only (re)open the URI when it wasn't previously opened, or when it was
closed by an earlier call.

This way, we can call `OSSL_STORE_open_ex()` in `by_store_ctrl_ex()`, and
get to see possible errors when the URI is loaded.

This assumes that if the URI could be opened once, it can be opened
again.

Fixes #27461
Alternative to #27507

levitte commented:
Side note: for anyone caring about the source format checker, there was an interesting false positive, see https://github.com/openssl/openssl/actions/runs/14751802024/job/41410783642?pr=27529 (@DDvO, @nhorman)

mattcaswell commented on crypto/x509/by_store.c:
```diff
@@ -115,27 +135,54 @@ static int by_store_ctrl_ex(X509_LOOKUP *ctx, int cmd, const char *argp,
-            if (uris == NULL) {
-                uris = sk_OPENSSL_STRING_new_null();
-                X509_LOOKUP_set_method_data(ctx, uris);
+            memset(data, 0, sizeof(*data));
```
Why not just use zalloc in the first place?

levitte replied:
Oh, 'cause I forget it exists...  fixing.

mattcaswell commented on crypto/x509/by_store.c:
```diff
@@ -115,27 +135,54 @@ static int by_store_ctrl_ex(X509_LOOKUP *ctx, int cmd, const char *argp,
+            }
+
+            if (stores == NULL) {
+                stores = sk_CACHED_STORE_new_null();
```
Check for NULL?

mattcaswell commented on crypto/x509/by_store.c:
```diff
@@ -152,13 +199,13 @@ static int by_store(X509_LOOKUP *ctx, X509_LOOKUP_TYPE type,
                     OSSL_LIB_CTX *libctx, const char *propq)
 {
-    STACK_OF(OPENSSL_STRING) *uris = X509_LOOKUP_get_method_data(ctx);
+    STACK_OF(CACHED_STORE) *uris = X509_LOOKUP_get_method_data(ctx);
```
This variable name no longer seems appropriate

levitte replied:
Good catch, thanks

mattcaswell reviewed (changes requested):
Not sure I really understand why this fixes #27461 which seems to be about absolute paths for the file scheme...or is it just the silent error part of #27461?

DDvO commented on crypto/x509/by_store.c:
```diff
@@ -152,13 +199,13 @@ static int by_store(X509_LOOKUP *ctx, X509_LOOKUP_TYPE type,
+    STACK_OF(CACHED_STORE) *uris = X509_LOOKUP_get_method_data(ctx);
     int i;
     int ok = 0;
 
-    for (i = 0; i < sk_OPENSSL_STRING_num(uris); i++) {
-        ok = cache_objects(ctx, sk_OPENSSL_STRING_value(uris, i), criterion,
-                           1 /* depth */, libctx, propq);
+    for (i = 0; i < sk_CACHED_STORE_num(uris); i++) {
+        ok = cache_objects(ctx, sk_CACHED_STORE_value(uris, i), criterion,
+                           1 /* depth */);
```
`uris` -> `stores`

levitte replied:
Yup

DDvO commented on crypto/x509/by_store.c:
```diff
@@ -91,21 +108,24 @@ static int cache_objects(X509_LOOKUP *lctx, const char *uri,
 }
 
 
 /* Because OPENSSL_free is a macro and for C type match */
```
The comment `/* Because OPENSSL_free is a macro and for C type match */`
should be removed now because it is no more to the point:
The new function does more than deallocation, it also closes the store.

DDvO commented on crypto/x509/by_store.c:
```diff
@@ -115,27 +135,54 @@ static int by_store_ctrl_ex(X509_LOOKUP *ctx, int cmd, const char *argp,
-            STACK_OF(OPENSSL_STRING) *uris = X509_LOOKUP_get_method_data(ctx);
-            char *data = OPENSSL_strdup(argp);
+            STACK_OF(CACHED_STORE) *stores = X509_LOOKUP_get_method_data(ctx);
+            CACHED_STORE *data = OPENSSL_malloc(sizeof(*data));
```
```suggestion
            CACHED_STORE *store = OPENSSL_malloc(sizeof(*store));
```

DDvO commented on crypto/x509/by_store.c:
```diff
@@ -115,27 +135,54 @@ static int by_store_ctrl_ex(X509_LOOKUP *ctx, int cmd, const char *argp,
+            if (data->ctx == NULL
+                || (propq != NULL && data->propq == NULL)
+                || data->uri == NULL) {
+                OSSL_STORE_close(data->ctx);
+                OPENSSL_free(data->propq);
+                OPENSSL_free(data->uri);
+                OPENSSL_free(data);
```
```suggestion
                close_free_store(store);
```

DDvO commented on crypto/x509/by_store.c:
```diff
@@ -91,21 +108,24 @@ static int cache_objects(X509_LOOKUP *lctx, const char *uri,
-    OPENSSL_free(data);
+    OSSL_STORE_close(store->ctx);
+    OPENSSL_free(store->uri);
+    OPENSSL_free(store->propq);
```
thereafter: 
`OPENSSL_free(store);`

levitte replied:
Thanks, fixed

DDvO commented on crypto/x509/by_store.c:
```diff
@@ -91,21 +108,24 @@ static int cache_objects(X509_LOOKUP *lctx, const char *uri,
 
 /* Because OPENSSL_free is a macro and for C type match */
-static void free_uri(OPENSSL_STRING data)
+static void close_store(CACHED_STORE *store)
```
```suggestion
static void close_free_store(CACHED_STORE *store)
```

levitte replied:
I just went for `free_store`

DDvO commented on crypto/x509/by_store.c:
```diff
@@ -115,27 +135,54 @@ static int by_store_ctrl_ex(X509_LOOKUP *ctx, int cmd, const char *argp,
             }
-            if (sk_OPENSSL_STRING_push(uris, data) <= 0) {
+            if (sk_CACHED_STORE_push(stores, data) <= 0) {
+                OSSL_STORE_close(data->ctx);
+                OPENSSL_free(data->propq);
+                OPENSSL_free(data->uri);
                 OPENSSL_free(data);
```
```suggestion
                close_free_store(store);
```

DDvO commented on crypto/x509/by_store.c:
```diff
@@ -115,27 +135,54 @@ static int by_store_ctrl_ex(X509_LOOKUP *ctx, int cmd, const char *argp,
+            memset(data, 0, sizeof(*data));
+            data->uri = OPENSSL_strdup(argp);
+            data->libctx = libctx;
+            if (propq != NULL)
+                data->propq = OPENSSL_strdup(propq);
```
Why dup'ing the string here, which is not done in `cache_objects()`?
After removing that, zero'ing the struct fields (e.g., via `memset()`) will no more be needed, as well as the below check on `data->propq`.

levitte replied:
Ah, I haven't really finished...  `by_store_subject_ex()` really shouldn't be a thing at all with these changes.  It's designed to override the libctx and propq that was passed through `by_store_ctrl_ex()`, which just seems *very* *wrong*.

levitte replied:
But, coming back to the question of duplication.  I duplicate here for exactly the same reason that we already duplicate the URI.  I hope this is reasonable at this point.

DDvO commented:
> Not sure I really understand why this fixes #27461 which seems to be about absolute paths for the file scheme...or is it just the silent error part of #27461?

@mattcaswell this new PR is about the silent error part.
The relative path extension has already been done.

DDvO commented:
> Fixes #27461 Alternative to #27507

On one hand, it provides a more thorough fix,
on the other, it does not contain fixes for related minor bugs.
One of those cause failure of the new `verify` test case "CAstore file://path":
https://github.com/openssl/openssl/actions/runs/14751802017/job/41410786941?pr=27529

levitte commented:
> ... it does not contain fixes for related minor bugs. One of those causes failure of the new `verify` test case "CAstore file://path": https://github.com/openssl/openssl/actions/runs/14751802017/job/41410786941?pr=27529

That's because `file://path` is a URI with the authority `path` and no path.

levitte commented:
> > ... it does not contain fixes for related minor bugs. One of those causes failure of the new `verify` test case "CAstore file://path": https://github.com/openssl/openssl/actions/runs/14751802017/job/41410786941?pr=27529
> 
> That's because `file://path` is a URI with the authority `path` and no path. Will fix that check.

... or huh???  hold on, something's weird with that one

levitte commented:
Ah, I see.  Yeah, please read RFC 8089, `file://C:/what/ever.pem` is non-standard, it should be `file:///C:/what/ever.pem`, i.e. the full URI path part should start with `/C:`

levitte commented:
> > Not sure I really understand why this fixes #27461 which seems to be about absolute paths for the file scheme...or is it just the silent error part of #27461?
> 
> @mattcaswell this new PR is about the silent error part. The relative path extension has already been done.

Yeah.  It's unfortunate that it got treated in the same issue, but quite frankly, this is at a point where raising a new issue "just 'cause" seems like unnecessary churn.

DDvO commented:
> Ah, I see. Yeah, please read RFC 8089, `file://C:/what/ever.pem` is non-standard, it should be `file:///C:/what/ever.pem`, i.e. the full URI path part should start with `/C:`

So it turns out that the Perl function ` abs_path()` it wrong, since it did not put a `/` before the drive letter.

levitte commented:
> > Ah, I see. Yeah, please read RFC 8089, `file://C:/what/ever.pem` is non-standard, it should be `file:///C:/what/ever.pem`, i.e. the full URI path part should start with `/C:`
> 
> So it turns out that the Perl function `abs_path()` is wrong, since it did not put a `/` before the drive letter.

It's perfectly functional, but requires a small tweak when used for a `file:` URI

levitte commented:
> BTW, while fixing `25-test_verify.t`, could you please also add a test case for a relative cert file name that does not have a scheme-like prefix, e.g.
> 
> ```
> my $relative_file = "cert.pem";
> copy($rootcert, $relative_file);
> ok(vfy_root("-CAstore", $relative_file), "CAstore relative_filename");
> ```

But...  we already have tests for relative paths!

t8m commented on crypto/x509/by_store.c:
```diff
@@ -91,21 +108,24 @@ static int cache_objects(X509_LOOKUP *lctx, const char *uri,
+static void free_store(CACHED_STORE *store)
 {
-    OPENSSL_free(data);
+    OSSL_STORE_close(store->ctx);
```
Please add `if (store == NULL) return;` to make it similar to other free functions.

levitte replied:
Fixed

t8m reviewed (changes requested):
One minor thing

levitte commented:
@mattcaswell, please re-approve.

DDvO commented:
> > BTW, while fixing `25-test_verify.t`, could you please also add a test case for a relative cert file name that does not have a scheme-like prefix, e.g.
> > ```
> > my $relative_file = "cert.pem";
> > copy($rootcert, $relative_file);
> > ok(vfy_root("-CAstore", $relative_file), "CAstore relative_filename");
> > ```
> 
> But... we already have tests for relative paths!

Yes, #27482 recently added such a test, from the `storeutil` perspective.
Here it would be in addition with the `verify` app.

With Windoze there is one further weird case, presumably inherited from CP/M: 
relative file pathnames with a drive letter, e.g., `file:D:cert.pem`.

I can add this case also in #27507, after I rebase on this PR when it got merged tomorrow.

levitte commented:
> > > BTW, while fixing `25-test_verify.t`, could you please also add a test case for a relative cert file name that does not have a scheme-like prefix, e.g.
> > > ```
> > > my $relative_file = "cert.pem";
> > > copy($rootcert, $relative_file);
> > > ok(vfy_root("-CAstore", $relative_file), "CAstore relative_filename");
> > > ```
> > 
> > 
> > But... we already have tests for relative paths!
> 
> Yes, #27482 recently added such a test, from the `storeutil` perspective. Here it would be in addition with the `verify` app.

You misunderstand.  This line, which is already present, is a test of a relative path, on Unix-like platforms:

```
ok(vfy_root("-CAstore", "file:".$foo_file), "CAstore file:foo:file");
```

However, yeah, that one is Unix only, and we should probably have a similar test without a "device"

levitte commented:
So I added an extra relative test, which means that all approvals need to be redone.  @mattcaswell @t8m @DDvO

DDvO commented:
> So I added an extra relative test,

Since you did not use a fixup commit it is hard to see what the actual change is.
I cannot spot a new test for relative file pathname, but just noticed that you meanwhile skip the two test cases involving `$foo_file` under Windows - I presume you did this because colons seem not allowed under Windows.

What I find a little problematic here is that since that `verify` parameter, allows for a URI,
the `CAstore foo:file` test case can make sense under Windows, namely if there is a provider handling the scheme `foo`.

DDvO commented:
> > Yes, #27482 recently added such a test, from the `storeutil` perspective. Here it would be in addition with the `verify` app.
> 
> You misunderstand. This line, which is already present, is a test of a relative path, on Unix-like platforms:
> 
> ```
> ok(vfy_root("-CAstore", "file:".$foo_file), "CAstore file:foo:file");
> ```
> 

This is not what I meant (note the PR number I gave), but the case
```
                    # Check relaxed 'file' scheme implementation
                    ok(run(app([@storeutl, "-noout", "-passin",
                                "pass:password", to_file_uri($_)])));
```
modified by your recent #27482, which now positively tests relative file names with.`storeutl`.

DDvO commented:
> With Windoze there is one further weird case, presumably inherited from CP/M: relative file pathnames with a drive letter, e.g., `file:D:cert.pem`.
> 
> I can add this case also in #27507, after I rebase on this PR when it got merged tomorrow.

Where and when to cover this special case?
Can you do here?

levitte commented:
> > So I added an extra relative test,
> 
> Since you did not use a fixup commit it is hard to see what the actual change is. I cannot spot a new test for relative file pathname,

The latest modification is visible if you know to click the `force-pushed` link:

![foo](https://github.com/user-attachments/assets/7b6bad15-67ce-4ff3-86c2-a06aa375fcad)

> but just noticed that you meanwhile skip the two test cases involving `$foo_file` under Windows - I presume you did this because colons are (plausibly) not allowed under Windows.

Because on Windows, `foo:` is a symbolic device, which is likely not to exist.

However, on Unix, `file:foo:cert.pem` is a `file:` URI with the *relative path* `foo:cert.pem`.  Like I said, it was already there.  But just to make sure, I also added a similar test with `$foo_file = "cert.pem";`, adding a relative path test that works on any platform.

> What I find a little problematic here is that since that `verify` parameter allows for a URI, the `CAstore foo:file` test case can make sense even under Windows, namely if there is a provider handling the scheme `foo`.

Yes, that's perfectly possible, if you use a provider that provides an implementation for that scheme.  That's not the case in these tests.  And yeah, it does mean that you need to *know your providers* (a mantra I've repeated so many times during the development of OpenSSL 3.0).  But also, if you study `OSSL_STORE_open_ex()`, you'll see that it tries using the `file:` scheme implementation first, and if the given URI has what looks like an explicit scheme, that one will be tried second.

levitte commented:
> > With Windoze there is one further weird case, presumably inherited from CP/M: relative file pathnames with a drive letter, e.g., `file:D:cert.pem`.
> > I can add this case also in #27507, after I rebase on this PR when it got merged tomorrow.
> 
> Where and when to cover this special case? Can you add a test case here?

Have you actually seen that special case live?  According to RFC 8089, it's not a valid *standard* file: URI.

DDvO commented:
> The latest modification is visible if you know to click the `force-pushed` link:

Good to learn!

> > but just noticed that you meanwhile skip the two test cases involving `$foo_file` under Windows - I presume you did this because colons are (plausibly) not allowed under Windows.
> 
> Because on Windows, `foo:` is a symbolic device, which is likely not to exist.

Yep.
 
> However, on Unix, `file:foo:cert.pem` is a `file:` URI with the _relative path_ `foo:cert.pem`. Like I said, it was already there. But just to make sure, I also added a similar test with `$foo_file = "cert.pem";`, adding a relative path test that works on any platform.

Nice.
 
> Yes, that's perfectly possible, if you use a provider that provides an implementation for that scheme. That's not the case in these tests. And yeah, it does mean that you need to _know your providers_ (a mantra I've repeated so many times during the development of OpenSSL 3.0).

Yes, were unlikely that we will ever have a provider here that handles `foo`.

> But also, if you study `OSSL_STORE_open_ex()`, you'll see that it tries using the `file:` scheme implementation first, and if the given URI has what looks like an explicit scheme, that one will be tried second.

Yes - and I admire the precision of your wording here and in the comment in `store_lib.c`: "looks like a scheme"
because the `strchr(scheme_copy, ':')` being used there constitutes an actually too weak check.
Instead, we should stop taking the prefix as a scheme at the first non-suitable character.
In #27507, I've been fixing the very same cute bug already for `OSSL_parse_url()`, and just have added a fix also for `OSSL_STORE_open_ex()`.

DDvO commented:
> > > With Windoze there is one further weird case, presumably inherited from CP/M: relative file pathnames with a drive letter, e.g., `file:D:cert.pem`.
> > > I can add this case also in #27507, after I rebase on this PR when it got merged tomorrow.
> > 
> > Where and when to cover this special case? Can you add a test case here?
> 
> Have you actually seen that special case live?

Yes - I clearly remember getting into this pitfall several times a couple of decades back when I was forced more often to use Windows.
A file pathname like `D:file` feels like an absolute path due to the drive letter in front, but in fact is a relative one (relative to something like the cwd on that drive).

> According to RFC 8089, it's not a valid _standard_ file: URI, and I'm unwilling to relax our implementation that far.

Yeah, while the RFC does not support that, technically it is possible on a Windows system.
I agree not to add special support for that, while I presume that it just works, for which I continue suggesting to add a test case.

> `file:/D:cert.pem`, however, is valid and should be recognised (if not, _please raise a separate issue for it_).

I'm not entirely sure this works with OpenSSL and I - gladly - cannot check directly because I don't have a Windows system at hand ;-)
The only thing I could do is to add further test cases for those Windows peculiarities and then wait and see if CI fails for them.

levitte commented:
BTW, I remembered it wrong.  Because of the change from #27482, `file:D:whatever.pem` should be accepted.  It's only `file://D:whatever.pem` that isn't (it should be `file:///D:whatever.pem`)

DDvO commented:
> BTW, I remembered it wrong. Because of the change from #27482, `file:D:whatever.pem` should be accepted. It's only `file://D:whatever.pem` that isn't (it should be `file:///D:whatever.pem`)

IIRC, #27507 is going to add support even for that.

levitte commented:
Anyway.  From my perspective, this PR is done.  Other issues deserve their own PR

DDvO commented:
I've just decided for myself that I won't handle anything further regarding Windoze peculiarities (unless I get really forced to do so).

I pointed out above some potentially further minor issue in that direction and now leave it to the rest of the team if anyone is up for adding any further tests and possibly fixes.

DDvO commented:
How about backporting this fix?

t8m reviewed (approved):
I am not sure about backporting this. Is there any potential for breaking things by backporting?

DDvO commented:
> I am not sure about backporting this. Is there any potential for breaking things by backporting?

In positive cases, there is no change except that cert stores opened earlier.
The point of this change is that now there is appropriate feedback in error cases opening and reading cert stores.

levitte commented:
I don't see why this couldn't be backported.  It should actually be possible to cherry-pick this PR to the release branches without too much trouble.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been  updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

levitte commented:
I was wrong re cherry-picking.

On 3.5 and 3.4, some of the tests fail because the relaxation of `file:` to accept relative paths when the scheme is explicit hasn't backported.
On 3.3 and older, there are code differences in crypto/x509/by_store.c that require an effort to get right, more than a simple cherry-pick allows.

So backporting will have to happen in separate PRs.

Unfortunately, I pushed to 3.5 before running local build and test.  I propose that fixing the tests is the most viable option, which is actually the backport effort needed for 3.5 and 3.4.

levitte commented:
Merged

master:
af5952d533b772ef8a3d7c666ed918acfc1dd911 (HEAD -> fix-27461) Drop "by store"'s by_store_subject_ex()
927debaf7b4e83e47d3f016808acab947cc50ea2 Add test_verify tests
0c48ee2bf513cbc2f1de2ff8bc11750e4b593620 Rework the "by store" X509_LOOKUP method to open the given URI early

3.5:
2e781d56a48b90faf86be3a93844a40d089c5477 (HEAD -> openssl-3.5) Drop "by store"'s by_store_subject_ex()
d268ec75cb39bf6ee17d49c712ad4cbbfbfdd7aa Add test_verify tests
295b5fd2e677531cc5ea25c38d2966dbbeb26271 Rework the "by store" X509_LOOKUP method to open the given URI early
(this turns out to need a fix, see #27549)

All other current release branches will have separate backport PRs

DDvO commented:
> So backporting will have to happen in separate PRs.

I just did for 3.4 in #27558.

> On 3.3 and older, there are code differences in crypto/x509/by_store.c that require an effort to get right, more than a simple cherry-pick allows.

I tried to come up with a respective backport PR, 
but as the conflicts are not straightforward to resolve, so I don't dare to do with my limited overview of OSSL_STORE.

levitte commented:
I've done the backports in:

- for 3.4: #27550 
- for 3.3, 3.2 and 3.0: #27551
